lib/core/refresh.js

```diff
@@ -0,0 +1,246 @@
+      }
+
+      // Refresh token before sending current request
+      await this.handleRefresh()
+
+      // Fetch updated token and add to current request
+      return this.getUpdatedRequestConfig(config)
+    })
+  }
+
+  reset () {
+    clearInterval(this.refreshInterval)
+  }
+
+  getUpdatedRequestConfig (config) {
+    config.headers[this.scheme.options.tokenName] = this.scheme.$auth.getToken(this.scheme.name)
+    return config
+  }
+
+  requestHasAuthorizationHeader (config) {
+    return !!config.headers.common[this.scheme.options.tokenName]
+  }
+}
+
+const TokenExpirationStatusEnum = Object.freeze({
+  UNKNOWN: 'UNKNOWN',
+  VALID: 'VALID',
+  EXPIRED: 'EXPIRED',
+  REFRESH_EXPIRED: 'REFRESH_EXPIRED'
+})
+
+export class TokenExpirationStatus {
+  constructor (scheme) {
+    this.$scheme = scheme
+  }
+
+  _calculateTokenStatus () {
+    const now = Date.now()
+    let tokenExpiresAt, refreshTokenExpiresAt
+
+    try {
+      tokenExpiresAt = this.$scheme._getTokenExpiration()
+      refreshTokenExpiresAt = this.$scheme._getRefreshTokenExpiration()
+
+      if (!tokenExpiresAt) {
+        return TokenExpirationStatusEnum.UNKNOWN
+      }
+    } catch (error) {
+      return TokenExpirationStatusEnum.UNKNOWN
+    }
+
+    // Give us some slack to help the token from expiring between validation and usage
+    const timeSlackMillis = 500
+    tokenExpiresAt -= timeSlackMillis
+    refreshTokenExpiresAt -= timeSlackMillis
+
+    // Token is still valid
+    if (now < tokenExpiresAt) {
+      return TokenExpirationStatusEnum.VALID
+    }
+
+    if (now > refreshTokenExpiresAt) {
+      return TokenExpirationStatusEnum.REFRESH_EXPIRED
+    }
+
+    return TokenExpirationStatusEnum.EXPIRED
+  }
+
+  setTokenExpiration (result = {}) {
+    const token = this.$scheme.$auth.getToken(this.$scheme.name)
+    let decodedToken, tokenExpiration
+    const _tokenIssuedAt = getProp(result, this.$scheme.options.issuedAt) || Date.now()
+    const _tokenTTL = getProp(result, this.$scheme.options.expiresIn) || this.$scheme.options.token.maxAge
+    const _tokenExpiresAt = getProp(result, this.$scheme.options.expiresAt) * 1000 || _tokenIssuedAt + (_tokenTTL * 1000)
+
+    try {
+      decodedToken = jwtDecode(token)
+      tokenExpiration = decodedToken.exp * 1000
+    } catch (error) {
+      // If the token is not jwt, we can't decode and refresh it, use _tokenExpiresAt value
+      tokenExpiration = _tokenExpiresAt || false
+
+      if (!(error instanceof InvalidTokenError)) {
+        throw error
+      }
+    }
+
+    // Set token expiration
+    this.$scheme._setTokenExpiration(tokenExpiration)
+
+    // Update status
+    this.syncStatus()
+
+    return tokenExpiration
+  }
+
+  setRefreshTokenExpiration () {
+    const refreshToken = this.$scheme.$auth.getRefreshToken(this.$scheme.name)
+    let decodedToken, refreshTokenExpiration
+    const _tokenIssuedAt = Date.now()
+    const _tokenTTL = this.$scheme.options.refreshToken.maxAge
```

Comment:
Isn't it best to make all timings consistent to ms?

Reply:
So you mean doing this? `const _tokenTTL = this.$scheme.options.refreshToken.maxAge + 1000`. I think that is better. But if you mean to change the option to expect ms, then I disagree.

Reply:
Makes sense using seconds too as `cookie.maxAge` can be a number and is more readable :) 

But then I'm worried about internal variables like `_tokenExpiresAt`, `_tokenTTL`, `refreshTokenExpiration`, `intervalDuration`, `_getTokenExpiration()`, ... that are in milliseconds and can be easily confused.

What if as a convention we assume all numeric timing units are either in milliseconds or seconds and add `ms` suffix (`refreshTokenExpiration` / ` to variables and options which are not? The reason I suggested ms was apparently most of the internal operations is with ms.

Reply:
Fixed in commit https://github.com/nuxt-community/auth-module/pull/548/commits/a7598cde31f147ab95d201b21b5f32e83dfc83ea